drivers/storage/portworx/component/px_resource_gateway.go

```diff
@@ -0,0 +1,444 @@
+)
+
+type pxResourceGateway struct {
+	isCreated  bool
+	k8sClient  client.Client
+	k8sVersion version.Version
+}
+
+func (p *pxResourceGateway) Name() string {
+	return PxResourceGatewayComponentName
+}
+
+func (p *pxResourceGateway) Priority() int32 {
+	return DefaultComponentPriority
+}
+
+func (p *pxResourceGateway) Initialize(
+	k8sClient client.Client,
+	k8sVersion version.Version,
+	_ *runtime.Scheme,
+	_ record.EventRecorder,
+) {
+	p.k8sClient = k8sClient
+	p.k8sVersion = k8sVersion
+}
+
+func (p *pxResourceGateway) IsPausedForMigration(cluster *corev1.StorageCluster) bool {
+	return util.ComponentsPausedForMigration(cluster)
+}
+
+func (p *pxResourceGateway) IsEnabled(cluster *corev1.StorageCluster) bool {
+	return cluster.Spec.PxResourceGateway != nil && cluster.Spec.PxResourceGateway.Enabled
+}
+
+func (p *pxResourceGateway) Reconcile(cluster *corev1.StorageCluster) error {
+	ownerRef := metav1.NewControllerRef(cluster, pxutil.StorageClusterKind())
+	if err := p.createServiceAccount(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+	if err := p.createClusterRole(); err != nil {
+		return err
+	}
+	if err := p.createClusterRoleBinding(cluster.Namespace); err != nil {
+		return err
+	}
+	if err := p.createDeployment(cluster, ownerRef); err != nil {
+		return err
+	}
+	if err := p.createService(cluster, ownerRef); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *pxResourceGateway) Delete(cluster *corev1.StorageCluster) error {
+	ownerRef := metav1.NewControllerRef(cluster, pxutil.StorageClusterKind())
+	if err := k8sutil.DeleteServiceAccount(p.k8sClient, PxResourceGatewayServiceAccountName, cluster.Namespace, *ownerRef); err != nil {
+		return err
+	}
+	if err := k8sutil.DeleteClusterRole(p.k8sClient, PxResourceGatewayClusterRoleName); err != nil {
+		return err
+	}
+	if err := k8sutil.DeleteClusterRoleBinding(p.k8sClient, PxResourceGatewayClusterRoleBindingName); err != nil {
+		return err
+	}
+	if err := k8sutil.DeleteDeployment(p.k8sClient, PxResourceGatewayDeploymentName, cluster.Namespace, *ownerRef); err != nil {
+		return err
+	}
+	p.MarkDeleted()
+	return nil
+}
+
+func (p *pxResourceGateway) MarkDeleted() {
+	p.isCreated = false
+}
+
+func (p *pxResourceGateway) createServiceAccount(
+	clusterNamespace string,
+	ownerRef *metav1.OwnerReference,
+) error {
+	return k8sutil.CreateOrUpdateServiceAccount(
+		p.k8sClient,
+		&v1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            PxResourceGatewayServiceAccountName,
+				Namespace:       clusterNamespace,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+		},
+		ownerRef,
+	)
+}
+
+func (p *pxResourceGateway) createClusterRole() error {
+	rules := []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{""},
+			Resources: []string{"configmaps"},
+			Verbs:     []string{"create", "get", "list", "patch", "update", "watch"},
+		},
+	}
+	return k8sutil.CreateOrUpdateClusterRole(
+		p.k8sClient,
+		&rbacv1.ClusterRole{
```

Comment:
Is there is a need to use cluster role instead of role? We should restrict the permissions to our components whenever possible. At least the new ones that we are adding now.

I believe we can create all the configmaps in the namespace of the storage cluster (where all px components are deployed). We should not even try to pollute other namespaces.